src/pyhf/constraints.py

```diff
@@ -125,62 +151,82 @@ def _precompute(self):
             )
-            self.poisson_rate_idc = tensorlib.astensor(
-                default_backend.tolist(poisson_rate_idc), dtype='int'
-            )
-            self.poisson_rate_fac = tensorlib.astensor(
-                default_backend.tolist(poisson_rate_fac), dtype='float'
+            factors = default_backend.reshape(
+                _poisson_rate_fac, (1, -1)
+            )  # (1, normals)
+            self._batched_factors = default_backend.tile(
+                factors, (self.batch_size or 1, 1)
             )
+
+            # self.parameter_helper.index_selection  is [ (batch, parslice) ] ]
+            access_field = None
+            for x in self.parameter_helper.index_selection:
+                access_field = (
+                    x
+                    if access_field is None
+                    else default_backend.concatenate([access_field, x], axis=1)
+                )
+            # access field is (nbatch, normals)
+            self._access_field = access_field
+
         else:
-            self.poisson_rate_idc, self.poisson_data, self.poisson_rate_fac = (
-                None,
-                None,
-                None,
-            )
+            self._poisson_data = None
+            self._access_field = None
+            self._batched_factors = None
+
+        self._precompute()
+        events.subscribe('tensorlib_changed')(self._precompute)
+
+    def _precompute(self):
+        if not self.parameter_helper.index_selection:
+            return
+        tensorlib, _ = get_backend()
+        self.poisson_data = tensorlib.astensor(self._poisson_data, dtype='int')
+        self.access_field = tensorlib.astensor(self._access_field, dtype='int')
+        self.batched_factors = tensorlib.astensor(self._batched_factors)
 
     def logpdf(self, auxdata, pars):
-        if self.poisson_data is None:
-            return 0
         tensorlib, _ = get_backend()
-        poisson_data = tensorlib.gather(auxdata, self.poisson_data)
-        poisson_rate_base = tensorlib.gather(pars, self.poisson_rate_idc)
-        poisson_factors = self.poisson_rate_fac
+        if not self.parameter_helper.index_selection:
+            return (
+                tensorlib.zeros(self.batch_size)
+                if self.batch_size is not None
+                else tensorlib.astensor(0.0)[0]
+            )
+        tensorlib, _ = get_backend()
 
-        poisson_rate = tensorlib.product(
-            tensorlib.stack([poisson_rate_base, poisson_factors]), axis=0
+        pars = tensorlib.astensor(pars)
+        if self.batch_size == 1 or self.batch_size is None:
+            batched_pars = tensorlib.reshape(
+                pars, (self.batch_size or 1,) + tensorlib.shape(pars)
+            )
+        else:
+            batched_pars = pars
+        flat_pars = tensorlib.reshape(batched_pars, (-1,))
+        nuispars = tensorlib.gather(flat_pars, self.access_field)
+
+        pois_rates = tensorlib.product(
```

Comment:
need a comment here for why the product/stack

Reply:
done